CommonClient.py

```diff
@@ -173,10 +174,82 @@ class CommonContext:
+                           f"names, attempt lookup with the game name. See f{self.lookup_type}_names docstring for "
+                           f"usage examples.")
+            name = self._store.get(key, f"Unknown {self.lookup_type} (ID: {key})")
+            return name if name is not None else f"Ambiguous {self.lookup_type} (ID: {key})"
+
+        def __setitem__(self, key, value):
+            if isinstance(key, int) and key in self._store:
+                # A `None` value signifies a duplicate item/location with the same id already exists. If anyone is
+                # trying to do an implicit name lookup, we can't be sure which game it is, so this will cause
+                # 'Ambiguous item/location' to be returned when `__getitem__` is called.
+                self._store[key] = None
+                return
+
+            self._store[key] = value
+
+        def __delitem__(self, key):
+            del self._store[key]
+
+        def __len__(self):
+            return len(self._store)
+
+        def __iter__(self):
+            return iter(self._store)
+
+        def __repr__(self):
+            return self._store.__repr__()
+
+        def update_game(self, game: str, lookup_table: typing.Dict[str, int]):
+            reverse_lookup_table = Utils.KeyedDefaultDict(lambda key: f"Unknown {self.lookup_type} (ID: {key})")
+            reverse_lookup_table.update({code: name for name, code in lookup_table.items()})
+            self._store[game] = collections.ChainMap(self._archipelago_lookup, reverse_lookup_table)
+            self.update(reverse_lookup_table)
+            if game == "Archipelago":
+                # Keep track of the Archipelago data package separately so if it gets updated in a custom datapackage,
+                # it updates in all chain maps automatically.
+                self._archipelago_lookup.clear()
+                self._archipelago_lookup.update(reverse_lookup_table)
+
     # data package
     # Contents in flux until connection to server is made, to download correct data for this multiworld.
-    item_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown item (ID:{code})')
-    location_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown location (ID:{code})')
+    item_names: ImplicitNameLookupDict
+    """A dictionary lookup of id -> name for items and updates automatically when a new data package arrives. If game 
+    name is omitted, it will attempt a lookup in all game packages.
+    
+    Returns "Unknown item" if an id does not exist in session data package.
+    Returns "Ambiguous item" if a game name is not provided and multiple items share the same id.
+    
+    Example usages::
+        name = item_names[game][id]  # Explicit id lookup with game name.
+        name = item_names[id]        # Implicit id lookup. May return "Ambiguous Item" if duplicate ids exist.
```

Comment:
I am opposed to this interface.

When I write this code:
```python
        x = self.item_names["game_name"]
        y = self.item_names[37]
        z = self.item_names["game_name"][37]
```
PyCharm, VSCode, mypy, pyright all have no idea what types any of these are.
And I doubt there is any way to add type annotations to help.

This hurts auto-complete, find references, type safety and all kinds of other IDE and tooling features.

---
If it had a separate interface for the game lookup that doesn't clash with the implicit lookup - something like this:
```python
    name = item_names.game(game_name)[id]
```
This would enable a much better IDE and tooling experience.

Reply:
I was trying to avoid having two different stores, but ultimately this might be the better interface. I'll play around with it a bit more when I have some free time tonight. ~~Not looking forward to trying to find all usages of `item_names` and `location_names` again lol.~~

Reply:
Another option:
If we formally say that we are deprecating the `item_names[id]` lookup, then I think the `item_names[game][id]` lookup interface is ok, and we can get good typing on it.
Type checkers and IDEs would report type errors for people using the `item_names[id]` lookup, but that's ok, because it's deprecated.

But if we are deprecating it, we should remove it from this docstring, and make it clear which parts of the code should be removed when we want to fully remove the deprecated functionality.
(And two different stores might make it easier to remove the deprecated functionality.)

Reply:
Good point on deprecation, that will most likely be the avenue I take.

Reply:
<details>
<summary>you can try this if you want</summary>

    class NameLookupDict(typing.Mapping[str, typing.Mapping[int, str]]):
        """ { game_name: { item_id: item_name } } """

        lookup_type: typing.Literal["item", "location"]
        _archipelago_lookup: typing.Dict[int, str]
        _game_store: typing.Dict[str, typing.Mapping[int, str]]
        _deprecated_store: typing.Dict[int, typing.Union[str, None]]

        def __init__(self, lookup_type: typing.Literal["item", "location"]):
            self.lookup_type = lookup_type
            self._archipelago_lookup = {}
            self._game_store = collections.defaultdict(
                lambda: Utils.KeyedDefaultDict(lambda key: f"Unknown {lookup_type} (ID {key})"))
            self._deprecated_store = {}

        def __contains__(self, key: object) -> bool:
            # remove isinstance check when deprecated lookup is removed
            if isinstance(key, str):
                return key in self._game_store

            logger.warning(f"Attempting to do an implicit lookup for an {self.lookup_type}. To avoid ambiguous id "
                           f"names, attempt lookup with the game name. See {self.lookup_type}_names docstring for "
                           f"usage examples.")
            return key in self._deprecated_store

        def __len__(self) -> int:
            return self._game_store.__len__()

        def __iter__(self) -> typing.Iterator[str]:
            return self._game_store.__iter__()

        def __getitem__(self, key: str) -> typing.Mapping[int, str]:
            # remove isinstance check when deprecated lookup is removed
            if isinstance(key, str):  # type: ignore [reportUnnecessaryIsInstance]
                return self._game_store[key]

            logger.warning(f"Attempting to do an implicit lookup for an {self.lookup_type}. To avoid ambiguous id "
                           f"names, attempt lookup with the game name. See {self.lookup_type}_names docstring for "
                           f"usage examples.")
            name = self._deprecated_store.get(key, f"Unknown {self.lookup_type} (ID: {key})")
            return name if name is not None else f"Ambiguous {self.lookup_type} (ID: {key})"

        def _update_deprecated_data(self, data: typing.Mapping[int, str]) -> None:
            for k, v in data.items():
                if k in self._deprecated_store:
                    # A `None` value signifies a duplicate item/location with the same id already exists. If anyone is
                    # trying to do an implicit name lookup, we can't be sure which game it is, so this will cause
                    # 'Ambiguous item/location' to be returned when `__getitem__` is called.
                    self._deprecated_store[k] = None
                else:
                    self._deprecated_store[k] = v

        def __repr__(self) -> str:
            return self._game_store.__repr__()

        def update_game(self, game: str, lookup_table: typing.Dict[str, int]) -> None:
            reverse_lookup_table = Utils.KeyedDefaultDict(lambda key: f"Unknown {self.lookup_type} (ID: {key})")
            reverse_lookup_table.update({code: name for name, code in lookup_table.items()})
            self._game_store[game] = collections.ChainMap(self._archipelago_lookup, reverse_lookup_table)
            self._update_deprecated_data(reverse_lookup_table)
            if game == "Archipelago":
                # Keep track of the Archipelago data package separately so if it gets updated in a custom datapackage,
                # it updates in all chain maps automatically.
                self._archipelago_lookup.clear()
                self._archipelago_lookup.update(reverse_lookup_table)

</details>

Reply:
I have pushed new changes and a slightly different API.